File: sdk-python/dephealth/checks/amqp.py
"""AMQP (RabbitMQ) health checker."""

from __future__ import annotations

from dephealth.checker import CheckAuthError, CheckConnectionRefusedError, CheckTimeoutError
from dephealth.dependency import Endpoint


class AMQPChecker:
    """Health check for an AMQP broker via connection."""

    def __init__(
        self,
        timeout: float = 5.0,
        url: str = "",
    ) -> None:
        self._timeout = timeout
        self._url = url

    async def check(self, endpoint: Endpoint) -> None:
        """Establish an AMQP connection and close it."""
        try:
            import aio_pika
        except ImportError:
            msg = "aio-pika is required for AMQP checker"
            raise ImportError(msg) from None

        url = self._url or f"amqp://{endpoint.host}:{endpoint.port}/"
        try:
            connection = await aio_pika.connect_robust(url, timeout=self._timeout)
            await connection.close()
        except TimeoutError as exc:
            msg = f"AMQP connection to {endpoint.host}:{endpoint.port} timed out"
            raise CheckTimeoutError(msg) from exc
        except OSError as e:
            msg = f"AMQP connection to {endpoint.host}:{endpoint.port} refused: {e}"
            raise CheckConnectionRefusedError(msg) from e
        except Exception as e:
            _classify_amqp_error(e, endpoint)
            raise

    def checker_type(self) -> str:
        """Return the checker type."""
        return "amqp"


def _classify_amqp_error(err: Exception, endpoint: Endpoint) -> None:
    """Re-raise AMQP auth errors as CheckAuthError."""
    msg = str(err)
    if "403" in msg or "ACCESS_REFUSED" in msg:
        raise CheckAuthError(f"AMQP auth error at {endpoint.host}:{endpoint.port}: {msg}") from err

File: sdk-go/dephealth/checks/amqp.go
package checks

import (
	"context"
	"fmt"
	"net"
	"strings"

	amqp "github.com/rabbitmq/amqp091-go"

	"github.com/BigKAA/topologymetrics/sdk-go/dephealth"
)

// AMQPOption configures the AMQPChecker.
type AMQPOption func(*AMQPChecker)

// AMQPChecker performs health checks by establishing an AMQP connection.
// Only standalone mode is supported: creates a new connection per check.
// The check succeeds if the AMQP connection can be established.
type AMQPChecker struct {
	url string // full AMQP URL (overrides endpoint-based URL)
}

// WithAMQPURL sets a custom AMQP URL for connections.
// If set, the endpoint host/port are ignored.
func WithAMQPURL(url string) AMQPOption {
	return func(c *AMQPChecker) {
		c.url = url
	}
}

// NewAMQPChecker creates a new AMQP health checker with the given options.
func NewAMQPChecker(opts ...AMQPOption) *AMQPChecker {
	c := &AMQPChecker{}
	for _, opt := range opts {
		opt(c)
	}
	return c
}

// Check establishes an AMQP connection and immediately closes it.
// Uses context for cancellation/timeout via a goroutine wrapper
// since amqp091-go does not natively support context.
func (c *AMQPChecker) Check(ctx context.Context, endpoint dephealth.Endpoint) error {
	url := c.url
	if url == "" {
		url = fmt.Sprintf("amqp://guest:guest@%s/", net.JoinHostPort(endpoint.Host, endpoint.Port))
	}

	type dialResult struct {
		conn *amqp.Connection
		err  error
	}

	ch := make(chan dialResult, 1)
	go func() {
		conn, err := amqp.Dial(url)
		ch <- dialResult{conn: conn, err: err}
	}()

	select {
	case <-ctx.Done():
		return fmt.Errorf("amqp dial %s: %w", endpoint.Host, ctx.Err())
	case res := <-ch:
		if res.err != nil {
			return classifyAMQPError(res.err, endpoint.Host)
		}
		_ = res.conn.Close()
		return nil
	}
}

// Type returns the dependency type for this checker.
func (c *AMQPChecker) Type() string {
	return string(dephealth.TypeAMQP)
}

// classifyAMQPError wraps AMQP errors with appropriate classification.
func classifyAMQPError(err error, host string) error {
	msg := err.Error()
	// AMQP 403 ACCESS_REFUSED indicates authentication/authorization failure.
	if strings.Contains(msg, "403") || strings.Contains(msg, "ACCESS_REFUSED") {
		return &dephealth.ClassifiedCheckError{
			Category: dephealth.StatusAuthError,
			Detail:   "auth_error",
			Cause:    fmt.Errorf("amqp dial %s: %w", host, err),
		}
	}
	return fmt.Errorf("amqp dial %s: %w", host, err)
}
